### ISSUE TITLE
[BUG] After writing four cards, the writer stops working

### ISSUE THREAD

**Issue by florecitu** [placiibo/placiibo-bugs#36]
I detected this bug since the previous version. After successfully writing four to five cards, for some reason, the writer stops working. You can tell the issue arises because it takes too long to detect the cards, when it usually is almost immediate. When you take these cards on which it took too long to detect to the Switch to see if they are detected are containing an amiibo, it does not contain a recognizable file.

1. Write 4-5 cards with the usual procedure. They work fine.
2. Try writing a card with a usual procedure.
3. It will take too long to actually detect the card and then it will appear as successfully written.
4. Take card to Switch and is not detected as containing an amiibo

**Expected behaviour**
Card should be written with files detected as an amiibo and should be detected as an amiibo by Switch.

**Actual behaviour**
Cards written incorrectly and not detected as an amiibo.

**Reproducibility :** 
This usually happens after writing 4-5 cards after updating to a new version.

**iOS version**:
13.5.1

**iOS device**: 
iPhone 11, 64 gb

**Jailbroken**: 
Not jailbroken. 

**placiibo version + build**: 
0.2.0 (403)

**Comment by sunnythemagnificent** [comment]
I am having the same issue. I was able to write ONE coin that works correctly and is detected, but as of now it constantly says "an error has occurred" when trying to write to an NFC coin, and takes forever to recognize it. I even used the same BIN file that worked earlier. The only difference is I got a new phone (upgraded from an Iphone 7 to an Iphone SE. 

IOS: 
13.5.1 

IOS device: 
Iphone SE, 128 GB

Placibo Version: 
0.8.0 (403.1) 

NFC Tags Used:
https://www.amazon.com/gp/product/B07PFC2FH5/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1

**Comment by savannaleya** [comment]
I’m having the same issues as well. After writing one tag clicking the previous bin file it says error occurred. It also is taking a while to read the the nfc tag. 

 IOS : 13.5.1
IOS device : iPhone XR
Placiibo Version : 0.8.0 (403.1)
![ADC322C4-F24B-4B2B-B6A6-589F64440C10](https://user-images.githubusercontent.com/67885433/86550551-4d8dbe80-bf08-11ea-9128-43d7c8e1e32c.png)

**Comment by placiibo** [comment]
I can't reproduce this issue, could you please send me the logs by using the email function in the app so I can take a lookg at what's going on? Tagging @sunnythemagnificent and @savannaleya as well.

**Comment by placiibo** [comment]
@florecitu I think you were the one that send me logs? They show the following:

```
16:16:11.188 ☢️ Session invalidated with error: Session invalidated by user
16:18:49.315 ☢️ Session invalidated with error: Session invalidated by user
16:19:41.262 ☢️ Session invalidated with error: Session invalidated by user
16:20:28.414 ☢️ Session invalidated with error: Session invalidated by user
16:21:09.573 ☢️ Error during write: Tag connection lost
16:21:28.376 ☢️ Session invalidated with error: Session timeout
16:22:00.281 ☢️ Error during write: Tag connection lost
16:22:19.427 ☢️ Session invalidated with error: Session timeout
16:23:54.436 ☢️ Error during connect: Stack Error
16:23:54.442 ☢️ Session invalidated with error: Session invalidated by user
16:24:00.832 ☢️ Error during write: Tag connection lost
16:24:19.664 ☢️ Session invalidated with error: Session timeout
12:46:55.908 ☢️ Error during write: Tag connection lost
12:47:12.415 ☢️ Session invalidated with error: Session invalidated by user
12:47:40.558 ☢️ Error during write: Tag connection lost
12:47:59.439 ☢️ Session invalidated with error: Session timeout
12:53:46.596 ☢️ Error during connect: Stack Error
12:53:46.603 ☢️ Session invalidated with error: Session invalidated by user
12:53:52.600 ☢️ Error during connect: Stack Error
12:53:52.607 ☢️ Session invalidated with error: Session invalidated by user
12:59:04.218 ☢️ Error during write: Tag connection lost
12:59:11.799 ☢️ Session invalidated with error: Session invalidated by user
12:59:17.125 ☢️ Error during write: Tag connection lost
12:59:24.806 ☢️ Session invalidated with error: Session invalidated by user
12:59:38.168 ☢️ Error during write: Tag connection lost
12:59:44.806 ☢️ Session invalidated with error: Session invalidated by user
13:00:47.613 ☢️ Error during write: Tag connection lost
13:01:07.308 ☢️ Session invalidated with error: Session timeout
13:05:34.586 ☢️ Error during write: Tag connection lost
13:05:36.703 ☢️ Session invalidated with error: Session invalidated by user
```
Session invalidated by user = you cancelling
Stack Error = an error writing to the tag but since these only happen miliseconds before cancelling it seems it's a result of cancellation 
Session timeout = the phone has looked for the tag the maximum amount of time allowed by Apple but couldn't find any
The rest of the log is self-explanatory

To sum it up: it seems like there are issues connecting to the tag. Could be the tag quality, could be a case around the phone that interferes, could also be that it's not held against the phone at the right spot. Whatever the exact cause is, the phone has trouble connecting to the tag and maintaining a connection. 

The app itself can't change anything about it becaue it is dependent on what it's being told by the iPhone. It doesn't have direct access to the antenna, the iPhone's system acts as an intemediary. 

Please try improving the conditions or try it with a different tag (i.e. one with a better/bigger antenna)

**Comment by savannaleya** [comment]
I think this might have been the same problem with me, I realize it could have been my phonecase and didn’t even think about that! Sorry! I believe I’ve sent the logs but if not please let me know.

**Comment by placiibo** [comment]
I’ve received some more logs from @savannaleya and the original reporter and they show the same pattern. 

You can look at them yourself by going to your email’s ‘Send’ folder, opening up the email you’ve send and then tap on the log. 
In the example above I’ve stripped away the parts that are unrelated to this issue, but the entries that are relevant to this issue all end with ‘com.apple.corenfc.readersession.delegate’. 

Normally the ‘Stack error’ entries would cause worry because they indicate an issue during writing, but since these appear within a 100th of a second before manually cancelling the write session and because of how the log prioritises writing an entry they are directly related to tapping the cancel button. 

What essentially happens in those cases is that it started writing to the tag, it got cancelled, logged that there was an issue with writing and then scrambled to log the cancel action. 

Successful writes and the start of a successful write aren’t logged in that file because I tried to be privacy conscious and only log warnings and errors. 
This now makes it a bit harder to see when it goes from a positive state to a negative one. I might fine tune the logging in a future build to capture this transition and change the generic error message to one that more clearly reflects what the phone is reporting. ‘Tag connection lost’ is a bit more helpful to the user than ‘An error has occurred’.

**Comment by sunnythemagnificent** [comment]
So I attempted to do this without my case and did manage to get a "write" checkmark (I recorded my screen and will send the recording via email in a second) but still had the issue of it registering as "not an amiibo". 

These NFC coins have worked with this app before, so that's why I'm struggling to think it could be them. I have ordered other NFC stickers that are in the mail to try as well, but I am unsure what else to say. I even did this with my case (Otterbox Defender) removed in case this was an issue.

**Comment by sunnythemagnificent** [comment]
Also to clarify some things I tried to make sure weren't factors: 
- I put the coin as close to the NFC antenna as possible. 
- I kept the coin flat on my palm. 
- I used the same .bin file that had worked before. 
- I did not have a case on at all.

**Comment by placiibo** [comment]
Thanks for providing more logs. After careful assesment of the logs I've come to the conclusion that they all show the same pattern, pointing to an outside source for the issue while excluding the possibility that this is caused by a bug in the app itself. 

Due to the limited scope I have of what happens beyond the app I can only make educated guesses, but in order of likelihood one of the following, or a combination thereof,  is probably the cause for issues:

* User error: improving the conditions (e.g. different tag, different tag placement, without case/different case, holding the tag longer/long enough against the phone, etc. etc.) can help in this instance.

* Bug in iOS: Less likely, but there are [known cases](https://www.youtube.com/watch?v=lfnSCgqnTQQ) in the beta days of iOS 11 in particular with the iPhone 7 where a bug caused the phone to have difficulties sustaining a NFC connection (e.g. like during a write), so it's not unheard of. The recourse for this is to [file a bug report with Apple[(http://feedbackassistant.apple.com) and/or contact Apple Support.

* Something wrong with the phone('s antenna): Unlikely, but the recourse for this would be to contact Apple support. 

Since this isn't a bug of the app itself, the suggestions above is all that I can provide and with that I'll close this bug report. 

I will however improve the error message so it will mirror what the phone reports to the app as the issue, which should at least help out some people that have no idea what went wrong.

Feel free to make a new bug report if you experience any new issues.

P.s. The 'not an amiibo' occurence isn't tied to a bug either.
Depending on, if the system recognises it as an amiibo (which you can try out by trying to register an owner in the system menu) or not, it's either:
*  an amiibo not supported by the game
* one that is made with a corrupt binary (i.e. typically binary dumps from amiibo that have (owner) data, look for a version that mentions it's 'unregistered') or,
* an issue during writing that ties back to the remedies mentioned above

**Comment by sunnythemagnificent** [comment]
Okay, so I'm sorry to continue providing data but I thought this might be of interest: 

**1st:** I have determined that the reason the whole "not an amiibo thing" happens is because literally NOTHING is written on the tags at all when Placiibo attempts to write to them. I figured this out using a powersaves NFC reader that can read amiibos already (including fanmade ones), the result was the same when I put a "blank" NFC on the reader as when I put an NFC that I had tried to write with Placiibo. Let me know if you want me to attempt to get video of this, and I would be happy to send. **THIS EVEN INCLUDES WHEN I GET A "CHECKMARK" SAYING DATA HAS BEEN WRITTEN.**

**2nd:** It does not depend on what "type" of tag. I bought some NFC sticker tags, (different from the coins I had earlier) and these had the exact same error. 

**3rd:** I have removed cases, and made the environment as clean and easy as possible and held the NFC tag as long as I could, and still it would not write. 

**4th:** I doubt this is an antenna issue itself because if it were (in theory, this is just my rough understanding) other NFC things such as apple pay would not work. Considering my Iphone SE is less than a week old out of the box, and I successfully used Apple Pay today, the chances of it being a broken antenna are next to none. 

My guess is, it's either some sort of app issue you haven't figured out, or a specific OS issue. The reasoning for that being that we ALL (each one who has posted) have different models of the Iphone, so it doesn't seem to be a model-specific issue. 

I'll reach out to apple on my end and see if they have had any reports about such things.

**Comment by sunnythemagnificent** [comment]
And of course, after I write all this out, I try one more time and put it on the top right corner of my iPhone and it writes. LOOOOOOOOOOOOOOOOOOOOOOOOL. #nevermind

**Comment by placiibo** [comment]
@sunnythemagnificent 

> Okay, so I'm sorry to continue providing data but I thought this might be of interest:
> 
> **1st:** I have determined that the reason the whole "not an amiibo thing" happens is because literally NOTHING is written on the tags at all when Placiibo attempts to write to them. I figured this out using a powersaves NFC reader that can read amiibos already (including fanmade ones), the result was the same when I put a "blank" NFC on the reader as when I put an NFC that I had tried to write with Placiibo. Let me know if you want me to attempt to get video of this, and I would be happy to send. **THIS EVEN INCLUDES WHEN I GET A "CHECKMARK" SAYING DATA HAS BEEN WRITTEN.**
> 
> **2nd:** It does not depend on what "type" of tag. I bought some NFC sticker tags, (different from the coins I had earlier) and these had the exact same error.
> 
> **3rd:** I have removed cases, and made the environment as clean and easy as possible and held the NFC tag as long as I could, and still it would not write.
> 
> **4th:** I doubt this is an antenna issue itself because if it were (in theory, this is just my rough understanding) other NFC things such as apple pay would not work. Considering my Iphone SE is less than a week old out of the box, and I successfully used Apple Pay today, the chances of it being a broken antenna are next to none.
> 
> My guess is, it's either some sort of app issue you haven't figured out, or a specific OS issue. The reasoning for that being that we ALL (each one who has posted) have different models of the Iphone, so it doesn't seem to be a model-specific issue.
> 
> I'll reach out to apple on my end and see if they have had any reports about such things.

1st: With your permission I can take a look based on your unique identifier that you send along when you emailed logs, all log messages that originate from the app are kept encrypted on a server for 1 hour before deletion so if it was recent I can take a look, but I'm about 99% sure it got interrupted during the write session and the NFC reader you're using just doesn't know what to show when scanning a tag that doesn't have complete data on it. 

2nd & 3rd I'll skip because I've got nothing useful to say in regards to that. 

4th: For the record I also doubt something is wrong with the antenna, I think it's user error. It could however be a design flaw like in the very narrow beam the iPhone 7 (plus)'s antenna transmits. But, transmitting a small piece of data, like with Apple Pay is something entirely different than having a continues writing session.  

But like I said, I will more than happily go through the logs of your specific case, just need your permission before I pull them up.

**Comment by sunnythemagnificent** [comment]
@placiibo I mean if it will help the others you can absolutely do what you like, I think I figured out the "sweet spot" to hold the NFC tags to get it to work (I think) so hopefully that solves my problem, but IDK about the others. So sorry to be a bother. xD

**Comment by placiibo** [comment]
@sunnythemagnificent You're not a bother, no worries, you're just keeping me on my toes. 
So I've specifically looked into your case and it seems like, prior to finding the sweet spot, it lost connection with the tag mid-write. Which would explain why the NFC reader acts the same like it would with an empty tag because it doesn't know what to do with 'half a tag' so to say. 

On the upside, there's a fair chance, depending on where in the write process it got disturbed, that you can reuse the tag, the locking of the tag (which is necessary because the Switch refuses to read non-locked tags) happens towards the end of the write. 

[Here](https://we.tl/t-J8nxcX4HT0) you can see the log where it shows it was writing until it went wrong, that's what has been happening every time it went wrong. I'm glad you've been able to find a sweet spot. 

I'm sorry there's not more I can do at this point, but I hope I was able to shed some light on what's happening.

**Comment by becaraptor** [comment]
I am having this same issue. iPhone 7 and using the + version. I’ve managed to make about five or six amiibos and now I can’t get it to work. 

I’m having the exact same issues. Takes way too long. Says that it has done it and then the switch says it’s not an amiibo. With three different sets of nfc tags. Including the set where the six of them worked. 

The one thing I have noticed is that it seems to write. I cannot write over it and when my partner uses it on his reader it comes up as the right file. Based on your comments above I believe that the issue is with the closing of the tag. (I think you used different wording but now I’ve typed this comment I can’t go back up and see it) but I think it’s a finishing and making it readable issue rather than writing the data itself.
